### ISSUE TITLE
Select an account prompt showing duplicate "Personal Access Token" options when Visual Studio is also signed-in with GitHub

### ISSUE THREAD

**Issue by FBoucher** [git-ecosystem/git-credential-manager#1325]
### Feature description

I have multiple Git accounts.
When the popup Select Account shows up... I have the choice between "Personal Access token" and "Personal Access token".
There is no way to make any distinctions.

See screenshot

![image](https://github.com/git-ecosystem/git-credential-manager/assets/2404846/4bacd3a6-41ad-4525-803b-8779f04080dc)

**Comment by yuvve** [comment]
I have the same problem and I have noticed (after getting the same issue) the following: if you remove all git credentials from Windows Credential Manager, then GCM will prompt for login (meaning it has 0 credentials saved) - after logging in (ONCE) with my account and trying to use GCM, I get the same prompt as @FBoucher (just with my username in both options). The annoying part is that we are now being asked to choose a user every time we use GCM, since it believes we have multiple users' credentials saved. 

This started after installing the latest update (`2.41.0.windows.2`) of git.

**Comment by KirillOsenkov** [comment]
Same issue. Looks like I got updated to 2.2.1 released on July 5:
https://github.com/git-ecosystem/git-credential-manager/releases/tag/v2.2.1

Could it have regressed something?

**Comment by KirillOsenkov** [comment]
@mjcheetham could this be a regression? This dialog started showing up today. I work for Microsoft and have Microsoft and other orgs access if that matters. This dialog disrupts the workflow and the UX is poor (which PAT is which?). How do I view more info about these PATs and perhaps delete one of them?

**Comment by MSACATS** [comment]
I'm having this issue as well without multiple access tokens

![image](https://github.com/git-ecosystem/git-credential-manager/assets/24610295/e39995f1-5765-4bca-b5cd-b12cf6c98555)
 
I have the github plugin installed in VS; I'm guessing this is what's up (excerpted from cmdkey).  

    Target: GitHub for Visual Studio - https://github.com/
    Type: Generic
    User: MSACATS
    Local machine persistence

    Target: git:https://github.com
    Type: Generic
    User: MSACATS
    Local machine persistence

**Comment by mjcheetham** [comment]
Hello all! Thanks for your patience with this problem. We have now identified the cause of the issue, and [have a fix](https://github.com/git-ecosystem/git-credential-manager/pull/1328).

The recently added support for multiple accounts in GCM is erroneously detecting duplicate credentials when Visual Studio has a signed-in GitHub account. Selecting either of the options listed "Personal Access Token" will allow the Git operation you are running to proceed, but we appreciate this is a frustrating interruption.

We are working on getting out a new release of GCM, and subsequently Git for Windows (that bundles GCM) with a fix to this issue.

In the meantime, there are a few workarounds:

1. Downgrade your Git for Windows installation to [2.41.0.windows.1](https://github.com/git-for-windows/git/releases/tag/v2.41.0.windows.1).

2. Sign out/remove the GitHub account in Visual Studio, and [delete existing credentials from the Windows credential manager](https://support.microsoft.com/en-us/windows/accessing-credential-manager-1b5c916a-6a16-889f-8581-fc16e8165ac0). Specifically you'll need to delete the `https://github.com` and `GitHub for Visual Studio - https://github.com` entries but **_not_** the `git:https://github.com` entry:

![image](https://github.com/git-ecosystem/git-credential-manager/assets/5658207/194a7a0f-d129-4348-96f2-7f9a44a1d64c)

**Comment by FBoucher** [comment]
Great! That was quick looking forward to see to update :)

**Comment by mjcheetham** [comment]
We've now merged in the fix to `main`, and will start working on a hotfix release that includes this imminently.

I will update this issue once the release is complete.

**Comment by mjcheetham** [comment]
Hello all!

Git for Windows 2.41.0(3) has just been released in the last few minutes:

https://github.com/git-for-windows/git/releases/tag/v2.41.0.windows.3

This release includes GCM 2.2.2 that fixes this issue.

**Comment by tobiasweibel** [comment]
Hi all
I still get that dialog. I'm using mainly an enterprise github instance and accessing it from git CLI and Jetbrians products.
version: `git version 2.41.0.windows.3`
Anybody else can still reproduce it?

**Comment by MacKopes1** [comment]
Still getting the issue as well with git version 2.41.0.windows.3 whenever I fetch with vscode.

**Comment by RobCannon** [comment]
@mjcheetham , please re-open this issue.  This still occurs with version 2.2.2 of git-credential-manager.

**Comment by mjcheetham** [comment]
@RobCannon can you run `git credential-manager diagnose` and report back with the log file output please?

Do you know if you have a standalone version of GCM installed, alongside the version bundled with Git for Windows? You can check in Control Panel/Settings > Apps > Installed Apps to see if "Git Credential Manager" is listed separately.

**Comment by mjcheetham** [comment]
We have reconfirmed that Git for Windows 2.41.0(3) and GCM 2.2.2 should both correct the issue with the prompt.

What is likely to have happened is a misconfiguration about which version of GCM is actually being called by Git. The version bundled in Git for Windows is typically the default option to be called, however installing a standalone copy of GCM (using the installers from our releases page), means that Git gets configured to call the standalone version _always_, even if the latest Git for Windows bundles a newer version.

If you uninstall any standalone GCM (from control panel or settings) versions, and ensure that Git for Windows 2.41.0(3) is installed, you can confirm your configuration is correct by running:

```console
% git config --show-origin --get-all credential.helper
file:C:/Program Files/Git/etc/gitconfig manager
```

If you see any other entries this is likely the cause of the non-fix.

**Comment by mrleemon** [comment]
By running the above command I get:
```
%git config --show-origin --get-all credential.helper
file:C:/Program Files/Git/etc/gitconfig manager
file:C:/Users/xxxxxxxxxx/.gitconfig    wincred
```

**Comment by mjcheetham** [comment]
> How can I delete the bad one?

For your situation you can run the following to remove the `wincred` line:

```
git config --global --unset-all credential.helper
```

**Comment by mrleemon** [comment]
@mjcheetham I'm no longer seeing the account selection dialog. Thanks!

**Comment by tobiasweibel** [comment]
> > How can I delete the bad one?
> 
> For your situation you can run the following to remove the `wincred` line:
> 
> ```
> git config --global --unset-all credential.helper
> ```

Additionally, I had to open the Windows Credential Manager and delete the GitHub credential. When calling `git fetch` I had to re-enter the personal access token. But now I'm all set.

**Comment by beard7** [comment]
> Additionally, I had to open the Windows Credential Manager and delete the GitHub credential. When calling `git fetch` I had to re-enter the personal access token. But now I'm all set.

Yep. That's done the trick. Thanks

**Comment by aidtopia** [comment]
I'm still confused about how to resolve this.  After reading through this thread a few times, here's what I've tried:

1.  I've taken git updates and verified that my git version is 2.41.0.windows.3.

2.  I have signed out of Visual Studio.  I've never intentionally used git or GitHub through the IDE, but somehow Visual Studio does track my git status.

3.  I have issued `git config --global --unset-all credential.helper`.  Now `git config --show-origin --get-all credential.helper` lists only `file:C:/Program Files/Git/etc/gitconfig manager`.

I've seen at least one comment saying I need to "delete the GitHub" credential as if there's only one.  In Windows Credential Manager (which I didn't know was a thing until tonight), I see two possibly relevant entries:

*   `git:https://github.com`
*   `git:https://aidtopia@github.com`

Am I supposed to delete one or both of those?

The screenshot upthread seems to suggest I should keep those and instead delete `GitHub for Visual Studio - https://github.com` and `https://github.com/` (w/o the `git:` prefix and with the trailing slash), but I don't have either of those.

I currently have only one GitHub account, and it's the only one I've ever used from this machine.  So it's not clear to me why I even have two credentials for GitHub.

I vaguely recall having trouble setting up a personal access token way back when, so I don't relish having to figure out that process again.

What's my next step?

**Comment by RobCannon** [comment]
I am using WSL and I had to run this command from the WSL prompt to keep the dialog from popping up:

```
git config --global credential.helper "/mnt/c/Program\ Files/Git/etc/gitconfig.exe manager"
```

This is NOT what is documented in WSL docs here:
https://github.com/git-ecosystem/git-credential-manager/blob/main/docs/wsl.md

**Comment by ldennington** [comment]
@RobCannon - can you please open a WSL terminal, run the below command, and share the output here?

`git config --list --show-origin`

**Comment by mjcheetham** [comment]
@aidtopia

> I currently have only one GitHub account, and it's the only one I've ever used from this machine. So it's not clear to me why I even have two credentials for GitHub.

The likely reason you are seeing these two entries is because the `git:https://github.com` entry was probably made by Visual Studio, with the user name as "Personal Access Token", and GCM didn't want to overwrite this credential with its own (with username "aidtopia") in case it was a secondary account.

The easiest thing to do here for you is to delete _both_ `git:` entries and then re-authenticate from the command-line by doing a `git fetch` in an existing repos for example.

**Comment by aidtopia** [comment]
@mjcheetham 

tl;dr:  It seems to have worked. Thanks!

I'm still rather confused, though.

> probably made by Visual Studio, with the user name as "Personal Access Token"

Neither of the user names for me were "Personal Access Token."  In my case, the "Select an Account" dialog kept asking me to choose between "aidtopia" and "aidtopia".

> re-authenticate from the command-line by doing a git fetch

Well, the process starts on the command line, but it sends you to the browser, which eventually directs to a localhost address.  Between that blank page and the command line instructing me to complete authentication via the browser, I naturally concluded something had gone wrong.  But, no, that's apparently what success looks like!  Who knew?

**Comment by rothja** [comment]
I could not fix this (by upgrading git & removing my credentials and reauthenticating) until I did that plus uninstall VS Code. After removing my credentials, removing the credential manager, uninstalling VS Code, reinstalling VS Code, and then setting up my credential managers and git credentials again in git bash, it worked. Just in case this helps anyone...

**Comment by RobCannon** [comment]
Here is my git config from WSL:
```
file:/home/rob/.gitconfig       user.name=Rob Cannon
file:/home/rob/.gitconfig       user.email=redacted
file:/home/rob/.gitconfig       credential.helper=/mnt/c/Program\ Files/Git/mingw64/bin/git-credential-manager.exe
file:/home/rob/.gitconfig       fetch.prune=true
file:/home/rob/.gitconfig       push.default=simple
file:/home/rob/.gitconfig       init.defaultbranch=main
file:/home/rob/.gitconfig       help.autocorrect=1
file:/home/rob/.gitconfig       pull.rebase=false
file:/home/rob/.gitconfig       core.editor=code --wait
```

**Comment by RobCannon** [comment]
OK, now I am no longer getting the duplicate account dialog, but I get the authenticate dialog (which ends opens up a browser window to authenticate) every few minutes.

ARGHH!!!

**Comment by Djfe** [comment]
Even after running `git config --global --unset-all credential.helper`:
This was caused by there being two entries for me in 2.41.0(3):
I had to delete both at the same time
* git:https://github.com
* git:https://Djfe@github.com

The next time I needed credentials, I had to relog using the browser (once), that was it

So yeah, this was not fixed by an update, I had to do manual intervention and I'm likely not the only one still affected
(I'm not using WSL)

**Comment by RobCannon** [comment]
OK, everything has been working fine this morning.  I re-cloned my repos without specifying the userid in the url (as was suggested somewhere) and I have not seen a dialog since.  Hopefully that will continue through the whole day.

For anyone using WSL, yesterday (suddenly) WSL was unable to execute Windows executables and that resulted in a an error calling git-credential-manager.exe.

```
/mnt/c/Program\ Files/Git/mingw64/bin/git-credential-manager.exe get: 1: /mnt/c/Program\ Files/Git/mingw64/bin/git-credential-manager.exe: Exec format error
```
I found a fix here:
https://github.com/microsoft/WSL/issues/9887#issuecomment-1493370267

```
sudo tee /usr/lib/binfmt.d/WSLInterop.conf > /dev/null <<'EOF'
:WSLInterop:M::MZ::/init:PF
EOF
```

**Comment by mjcheetham** [comment]
> UPDATE: Ignore this. It was gone for 12 hours after doing this, and now the duplicate account and dialog is back. Not sure what to try now.

@rothja, please can you set the following Git config and when you get a repro of the repeated auth prompt, complete auth and then send over the resulting log file.

```shell
# Windows
git config --global credential.trace C:\Users\USERNAME\gcm.log
git config --global credential.traceMsAuth true

# Mac/Linux
git config --global credential.trace /tmp/gcm.log
git config --global credential.traceMsAuth true
```

You may want to glance over the file for any private information before sharing, for good measure.

To turn off the logging after, you can run the following:

```shell
git config --global --unset credential.trace
git config --global --unset credential.traceMsAuth
```

**Comment by RobCannon** [comment]
Another day, another problem.  Or rather back to the first problem.  I, too, am getting the duplicate account dialog again.

I am attaching my credential trace.
[gcm.log](https://github.com/git-ecosystem/git-credential-manager/files/12254642/gcm.log)

**Comment by RobCannon** [comment]
I just want to make sure that WSL should still have this setting:
```
>  git config --show-origin --get-all credential.helper
file:/home/rob/.gitconfig       /mnt/c/Program\ Files/Git/mingw64/bin/git-credential-manager.exe
```

It is different than what is set in my Windows prompt, which is:
```
>  git config --show-origin --get-all credential.helper
file:C:/Program Files/Git/etc/gitconfig manager
```

**Comment by Kalas777** [comment]
😆😁😁

**Comment by UTshion** [comment]
Hi, I am using VScode Remote Containers and had the same issue. 
However, thanks to the comments on this thread, I was able to prevent the user selection popup from appearing every time I perform a Git operation.

I think there might be others who use a similar environment to mine, so I would like to summarize here what I have done to deal with the situation. 
I hope my notes helps you.

---

### Environment

WSL2+Docker+VScode


### Versions

**[Host]**
Windows 11 Pro 22H2
WSL2 : 1.2.5.0
Docker Desktop : 4.22.0

**[Container]** 
Ubuntu 20.04.6 LTS
Git : 2.25.1
GCM : 2.2.2


## My operation

1. I ran `git config --show-origin --get-all credential.helper`, and found there were two credentials. Therefore, I also ran `git config --global --unset-all credential.helper` to delete the one.
  After that, I ran `git config --show-origin --get-all credential.helper` again and identified there was only one credential.


2. I signed out the Github account in VScode. Next, I opened **Windows Credential Manager**, and found there were some similar entries which contain the phrase "git" or "github".
   I deleted all of those credentials except git:https://github.com and Github - https://api.github.com/UTshion, but that didn't work!

3. **According to the comments above, I also deleted "git:https://github.com".** Fortunately, that worked! And after I re-authenticated my Git account in VScode once, I have never seen the poppup.
   During the re-authentication, as @mjcheetham mentioned, I was also faced the situation: the browser sent me to localhost address for requiring me to authenticate the account, but nothing was displayed. 
   It gives me somewhat hazy feelings, but I think I seem to have solve the problem for now, so I decided not to care...

---

**This is my first time writing such a long comment in English, so I apologize if it is unclear or incorrect.**

**Comment by RobCannon** [comment]
I followed the advice of @UTshion and I signed out of GitHub from VSCode, removed all of the Git and GitHub entries from Windows Credential manager and signed in again.  I have not had problems since, but I need to work through a whole business day or two before I can declare victory.

**Comment by RobCannon** [comment]
OMG, literally seconds after posting the above, I got the duplicate account dialog again!  It is showing with every git interaction again!!!

**Comment by RobCannon** [comment]
I am beginning to think that I have several extensions that all sign into GitHub and one of them is generating a different credential type and that is what causes the Duplicate Account dialog box.  Now to figure out which one is doing it...

**Comment by Kalas777** [comment]
Thanks github u are awesome

**Comment by ldennington** [comment]
@RobCannon - the following should force GCM to use the correct username for all GitHub repos:

`git config --global credential.https://github.com.username robcannon`

Hopefully that will resolve the issue for you.

**Comment by mjcheetham** [comment]
We've updated our docs on this issue for future reference: https://gh.io/gcm/multipleusers#tldr-tell-gcm-to-remember-which-account-to-use

@rothja this may also help solve your situation also.

**Comment by v-thepet** [comment]
YES. THANK YOU @ldennington - After many hours spent with GitHub Support and our company's IT team troubleshooting this issue, which started 7/13 for me, that line looks like it worked:

`git config --global credential.https://github.com.username v-thepet`

No more prompt on every remote action. 

I won't say so for sure just yet, as I thought the problem was solved a couple of times before and it came back.

**Comment by Kalas777** [comment]
Awesome works thank you😀😀😀

**Comment by FBoucher** [comment]
Feels like I should close this issue now, right?

**Comment by ldennington** [comment]
Yes, I believe we are good to close this - I will go ahead and do so.

**Comment by alexhiggins732** [comment]
git credential-manager github list returned no output for me.

My fix was simply to open the windows credential manager and modify the user name on each "Personal Access Token" account to "Personal Access Token 1" , "Personal Access Token 2", etc to identify the correct account. 

I then removed the duplicitous accounts in Windows credential manager and the issue went away.

**Comment by kostia-Etaplus** [comment]
I've deleted all the windows credentials records for github, logged out from 

> Hi, I am using VScode Remote Containers and had the same issue. However, thanks to the comments on this thread, I was able to prevent the user selection popup from appearing every time I perform a Git operation.
> 
> I think there might be others who use a similar environment to mine, so I would like to summarize here what I have done to deal with the situation. I hope my notes helps you.
> 
> ### Environment
> WSL2+Docker+VScode
> 
> ### Versions
> **[Host]** Windows 11 Pro 22H2 WSL2 : 1.2.5.0 Docker Desktop : 4.22.0
> 
> **[Container]** Ubuntu 20.04.6 LTS Git : 2.25.1 GCM : 2.2.2
> 
> ## My operation
> 1. I ran `git config --show-origin --get-all credential.helper`, and found there were two credentials. Therefore, I also ran `git config --global --unset-all credential.helper` to delete the one.
>    After that, I ran `git config --show-origin --get-all credential.helper` again and identified there was only one credential.
> 2. I signed out the Github account in VScode. Next, I opened **Windows Credential Manager**, and found there were some similar entries which contain the phrase "git" or "github".
>    I deleted all of those credentials except git:https://github.com and Github - https://api.github.com/UTshion, but that didn't work!
> 3. **According to the comments above, I also deleted "git:https://github.com".** Fortunately, that worked! And after I re-authenticated my Git account in VScode once, I have never seen the poppup.
>    During the re-authentication, as @aidtopia mentioned, I was also faced the situation: the browser sent me to localhost address for requiring me to authenticate the account, but nothing was displayed.
>    It gives me somewhat hazy feelings, but I think I seem to have solve the problem for now, so I decided not to care...
> 
> **This is my first time writing such a long comment in English, so I apologize if it is unclear or incorrect.**

Thanks. I've updated git bash first and then followed these instructions and finally it worked for me as well.

**Comment by Constantine-Ketskalo** [comment]
> Hi, I am using VScode Remote Containers and had the same issue. However, thanks to the comments on this thread, I was able to prevent the user selection popup from appearing every time I perform a Git operation.
> 
> I think there might be others who use a similar environment to mine, so I would like to summarize here what I have done to deal with the situation. I hope my notes helps you.
> 
> ### Environment
> WSL2+Docker+VScode
> 
> ### Versions
> **[Host]** Windows 11 Pro 22H2 WSL2 : 1.2.5.0 Docker Desktop : 4.22.0
> 
> **[Container]** Ubuntu 20.04.6 LTS Git : 2.25.1 GCM : 2.2.2
> 
> ## My operation
> 1. I ran `git config --show-origin --get-all credential.helper`, and found there were two credentials. Therefore, I also ran `git config --global --unset-all credential.helper` to delete the one.
>    After that, I ran `git config --show-origin --get-all credential.helper` again and identified there was only one credential.
> 2. I signed out the Github account in VScode. Next, I opened **Windows Credential Manager**, and found there were some similar entries which contain the phrase "git" or "github".
>    I deleted all of those credentials except git:https://github.com and Github - https://api.github.com/UTshion, but that didn't work!
> 3. **According to the comments above, I also deleted "git:https://github.com".** Fortunately, that worked! And after I re-authenticated my Git account in VScode once, I have never seen the poppup.
>    During the re-authentication, as @aidtopia mentioned, I was also faced the situation: the browser sent me to localhost address for requiring me to authenticate the account, but nothing was displayed.
>    It gives me somewhat hazy feelings, but I think I seem to have solve the problem for now, so I decided not to care...
> 
> **This is my first time writing such a long comment in English, so I apologize if it is unclear or incorrect.**

Thanks. I've updated git bash first and then followed these instructions and finally it worked for me as well. By the way I have both vs and vs code, so I've logged out from github in both apps just in case, even though I've been getting this duplicated account choice only from vs.
